### ISSUE TITLE
Merging two CellTag objects?

### ISSUE THREAD

**Issue by RobertKSuter** [morris-lab/CellTagR#4]
Hi, I've just realized that CellRanger Aggregate does not output a single bam file of aggregated samples. 

I have two different 10X runs (different samples) that should have overlapping Celltag combinations as each is from a common original Celltag transduction. 

The goal would be to build a single network across all three samples. Would you have any recommendations as the best way to do this?

 I could rerun cellranger count on both samples together for a single bam file, but I think I would then lose the source identity for any cells with a clashing 10X barcode. 

Thank you!

**Comment by sam-morris** [comment]
Hi Robert. One way to approach this is to merge your CellTag UMI matrix files after the last filtering step, here: https://github.com/morris-lab/CellTagR#4-apply-the-whitelisted-celltags-generated-from-assessment. Make sure you append your cell barcodes, so you know which run they came from, and so you don't lose source identity when you merge the files. Then proceed to the clone calling step, here: https://github.com/morris-lab/CellTagR#8-clone-calling

**Comment by RobertKSuter** [comment]
Hi Dr. Morris, thank you! I'll give this a shot - this would need to be done separately for each version, correct? 

I.e., for each version first append barcodes with a unique identifier:

```
bam.1.obj@bam.parse.rslt$v1$Cell.BC <- paste0("DMSO_", bam.1.obj@bam.parse.rslt$v1$Cell.BC)

bam.2.obj@bam.parse.rslt$v1$Cell.BC <- paste0("DMSO_", bam.2.obj@bam.parse.rslt$v1$Cell.BC)
```

 run object1 and object2 through to the last filtering step in parallel for v1. then, before clone-calling set:

`bam.1.obj@binary.mtx <- merge(bam.1.obj@binary.mtx, bam.2.obj@binary.mtx, ...)`

Thank you again for your help

**Comment by sam-morris** [comment]
Yes - so long as your identifier is unique, that should work. Good luck - Sam.

**Comment by RobertKSuter** [comment]
Great, thanks! 

Now that I've got both samples filtered, with unique cell barcode names, would you have a suggestion as how to best merge the two dgCMatrix slots? Using merge as the command above doesn't work, as it cannot coerce them to dataframes. 

By first converting each to individual matrices (instead of sparse matrices), I am able to then get them as a dataframe. However, running merge on these dataframes is very memory intensive (greater than 250gb of memory) , even if only merging on the intersection of the celltags in each sample.

```
A <- bam.NM002.obj@binary.mtx
B <- bam.DMSO.obj@binary.mtx

A.dataframe <- as.data.frame(as.matrix(A))
B.dataframe <- as.data.frame(as.matrix(B))

intersect(names(A.dataframe), names(B.dataframe))

c <- merge(A.dataframe, B.dataframe, by = intersect(names(A.dataframe), names(B.dataframe)))
```

Thank you again! - Rob

**Comment by sam-morris** [comment]
Hi Rob, @KaetheKong is working on a better solution now - we'll keep you updated.

**Comment by RobertKSuter** [comment]
Amazing! Thank you both!

If it would help, I would be happy to share the two objects as .Rds that I am trying to merge privately. All the best, Rob

**Comment by KaetheKong** [comment]
Hi Rob,

Thank you for using this tool!!

I just updated the package to enable multiple BAM file processing in the extraction stage as well as the README.md file. Following this pipeline should not need any manual merging. And hopefully not taking as much memory... Please give it a try and feel free to let us know if there are any issues with the new pipeline! 

Best,
Wenjun

**Comment by RobertKSuter** [comment]
Hi Wenjun,

Great, thank you! Will give it a shot.

All the best,
Rob

**Comment by KaetheKong** [comment]
Yes - based on the order of BAM file that you passed in, it will add Sample-1 before the barcodes of the first bam file cells and Sample-2 before the second.

**Comment by RobertKSuter** [comment]
Hi Wenjun,

For the barcodes.tsv, should I be appending Sample-1_, Sample-2_, etc. to the barcodes in the tsv file? Or will this mess up starcode in the collapsing step?

Thank you again!
Rob

**Comment by RobertKSuter** [comment]
Hi Wenjun,

I've now gotten to the post-starcode step, but I am getting the following error while running:

```
collapsed.rslt.dir <- "/int_starcode_rslt_dir"

list.files(collapsed.rslt.dir)

#Recount and generate collapsed matrix
bam.int.obj <- CellTagDataPostCollapsing(bam.int.obj, collapsed.rslt.file = list.files(collapsed.rslt.dir, full.names = T))
```
`Error in paste0("Processing ", curr.file.dir) : object 'curr.file.dir' not found`

I think this is because in the function CellTagDataPostCollapsing() curr.file.dir is called before it is assigned.

```
function (celltag.obj, collapsed.rslt.file) 
{
  ultimate.collapsing.df <- data.frame()
  for (i in 1:length(collapsed.rslt.file)) {
    print(paste0("Processing ", curr.file.dir))
    final.collapsing.df <- data.frame()
    curr.file.dir <- collapsed.rslt.file[i]
    collapsed <- read.table(curr.file.dir, sep = "\t", header = F, 
      stringsAsFactors = F)

```
Maybe should be...

```
function (celltag.obj, collapsed.rslt.file) 
{
  ultimate.collapsing.df <- data.frame()
  for (i in 1:length(collapsed.rslt.file)) {
    curr.file.dir <- collapsed.rslt.file[i]
    print(paste0("Processing ", curr.file.dir))
    final.collapsing.df <- data.frame()

    collapsed <- read.table(curr.file.dir, sep = "\t", header = F, 
      stringsAsFactors = F)

```
Do you think this is the problem?

Thank you!
Rob

**Comment by KaetheKong** [comment]
Hi Rob,

Thank you for pointing it out!

That is exactly where the problem is! I just updated that part.

Best,
Wenjun

**Comment by RobertKSuter** [comment]
Hi Wenjun, no problem! & Thank you!

I have one last question, would you have a quick suggestion on how I could color the force-directed graph network based on the two samples inputted? I.e. red for sample-1, blue for sample-2?

Thank you again!
Rob

**Comment by sam-morris** [comment]
Hi Rob - glad it worked out. I'll close this now but please feel free to open another issue for the graph networks. Best wishes, Sam.
